crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs

```diff
@@ -3227,6 +3227,8 @@ impl Analyzer<'_, '_> {
         }
 
         if let Type::Predicate(p) = ret_ty.normalize() {
+            self.ctx.is_type_predicate = true;
```

Comment:
Can we use `with_ctx` instead?

Reply:
That makes `pass_only_tests__pass_only__conformance__types__conditional__conditionalTypes2__3_ts
` fail with

(it fails on a function with a type predicate)

```rs
context: tried to validate an argument (at crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs:3206:25)
  │ context:
  │ lhs = instanceof  (object | Function);
  │ rhs = T#14#0 extends Function#0 ? T : never;
  │ lhs (normalized) = (object | Function#0);
  │ rhs (normalized) = T#14#0 extends Function#0 ? T : never; (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:587:73)
  │ context: tried to assign the true type of a conditional type to lhs (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:1089:22)
  │ context:
  │ lhs = (object | Function#0);
  │ rhs = T#14#0;
  │ lhs (normalized) = (object | Function#0);
  │ rhs (normalized) = T#14#0; (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:587:73)
  │ context: tried to assign a type to a union type (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:1985:67)
  │ WrongArgType {
  │     span: Span {
  │         lo: BytePos(
  │             1089,
  │         ),
  │         hi: BytePos(
  │             1094,
  │         ),
  │         ctxt: #14,
  │     },
  │     inner: Errors {
  │         span: Span {
  │             lo: BytePos(
  │                 1089,
  │             ),
  │             hi: BytePos(
  │                 1094,
  │             ),
  │             ctxt: #14,
  │         },
  │         errors: [
  │             context: tried to assign a type to a union (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:1951:26)
  │             context:
  │             lhs = object;
  │             rhs = T#14#0;
  │             lhs (normalized) = object;
  │             rhs (normalized) = T#14#0; (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:587:73)
  │             context: LHS (final): object;
  │             RHS (final): T#14#0; (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:1694:25)
  │             AssignFailed {
  │                 span: Span {
  │                     lo: BytePos(
  │                         1089,
  │                     ),
  │                     hi: BytePos(
  │                         1094,
  │                     ),
  │                     ctxt: #14,
  │                 },
  │                 cause: [],
  │             },
  │             context: tried to assign a type to a union (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:1951:26)
  │             context:
  │             lhs = Function#0;
  │             Member as {
  │                 apply(this: Function, thisArg: any, argArray?: any): any;
  │                 call(this: Function, thisArg: any, ...argArray: any[]): any;
  │                 bind(this: Function, thisArg: any, ...argArray: any[]): any;
  │                 toString(): string;
  │                 prototype: any;
  │                 readonly length: number;
  │                 arguments: any;
  │                 caller: Function;
  │             };
  │             rhs = T#14#0;
  │             lhs (normalized) = Function#0;
  │             Member as {
  │                 apply(this: Function, thisArg: any, argArray?: any): any;
  │                 call(this: Function, thisArg: any, ...argArray: any[]): any;
  │                 bind(this: Function, thisArg: any, ...argArray: any[]): any;
  │                 toString(): string;
  │                 prototype: any;
  │                 readonly length: number;
  │                 arguments: any;
  │                 caller: Function;
  │             };
  │             rhs (normalized) = T#14#0; (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:587:73)
  │             context: LHS (final): Function#0;
  │             Member as {
  │                 apply(this: Function, thisArg: any, argArray?: any): any;
  │                 call(this: Function, thisArg: any, ...argArray: any[]): any;
  │                 bind(this: Function, thisArg: any, ...argArray: any[]): any;
  │                 toString(): string;
  │                 prototype: any;
  │                 readonly length: number;
  │                 arguments: any;
  │                 caller: Function;
  │             };
  │             RHS (final): T#14#0; (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:1694:25)
  │             AssignFailed {
  │                 span: Span {
  │                     lo: BytePos(
  │                         1089,
  │                     ),
  │                     hi: BytePos(
  │                         1094,
  │                     ),
  │                     ctxt: #14,
  │                 },
  │                 cause: [],
  │             },
  │         ],
  │     },
  │ }
    ╭─[./a.ts:46:1]
 46 │     if (isFunction(value)) {
 47 │         toString1(value);
    ·                   ─────
 48 │         toString2(value);
    ╰────
TS2345

  × context: tried to validate an argument (at crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs:3206:25)
  │ context:
  │ lhs = instanceof  Function;
  │ rhs = T#14#0 extends Function#0 ? T : never;
  │ lhs (normalized) = Function#0;
  │ Member as {
  │     apply(this: Function, thisArg: any, argArray?: any): any;
  │     call(this: Function, thisArg: any, ...argArray: any[]): any;
  │     bind(this: Function, thisArg: any, ...argArray: any[]): any;
  │     toString(): string;
  │     prototype: any;
  │     readonly length: number;
  │     arguments: any;
  │     caller: Function;
  │ };
  │ rhs (normalized) = T#14#0 extends Function#0 ? T : never; (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:587:73)
  │ context: tried to assign the true type of a conditional type to lhs (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:1089:22)
  │ context:
  │ lhs = Function#0;
  │ Member as {
  │     apply(this: Function, thisArg: any, argArray?: any): any;
  │     call(this: Function, thisArg: any, ...argArray: any[]): any;
  │     bind(this: Function, thisArg: any, ...argArray: any[]): any;
  │     toString(): string;
  │     prototype: any;
  │     readonly length: number;
  │     arguments: any;
  │     caller: Function;
  │ };
  │ rhs = T#14#0;
  │ lhs (normalized) = Function#0;
  │ Member as {
  │     apply(this: Function, thisArg: any, argArray?: any): any;
  │     call(this: Function, thisArg: any, ...argArray: any[]): any;
  │     bind(this: Function, thisArg: any, ...argArray: any[]): any;
  │     toString(): string;
  │     prototype: any;
  │     readonly length: number;
  │     arguments: any;
  │     caller: Function;
  │ };
  │ rhs (normalized) = T#14#0; (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:587:73)
  │ context: LHS (final): Function#0;
  │ Member as {
  │     apply(this: Function, thisArg: any, argArray?: any): any;
  │     call(this: Function, thisArg: any, ...argArray: any[]): any;
  │     bind(this: Function, thisArg: any, ...argArray: any[]): any;
  │     toString(): string;
  │     prototype: any;
  │     readonly length: number;
  │     arguments: any;
  │     caller: Function;
  │ };
  │ RHS (final): T#14#0; (at crates/stc_ts_file_analyzer/src/analyzer/assign/mod.rs:1694:25)
  │ WrongArgType {
  │     span: Span {
  │         lo: BytePos(
  │             1115,
  │         ),
  │         hi: BytePos(
  │             1120,
  │         ),
  │         ctxt: #14,
  │     },
  │     inner: AssignFailed {
  │         span: Span {
  │             lo: BytePos(
  │                 1115,
  │             ),
  │             hi: BytePos(
  │                 1120,
  │             ),
  │             ctxt: #14,
  │         },
  │         cause: [],
  │     },
  │ }
    ╭─[./a.ts:47:1]
 47 │         toString1(value);
 48 │         toString2(value);
    ·                   ─────
 49 │     }
    ╰────
```
but I suppose that's because my solution isn't entirely correct? I can investigate more if we have to use `with_ctx`

Reply:
If so, I think it's one of

 - You didn't use `with_ctx` in some code. You have to wrap all operations on `self` with it.
 - It passes wrongly while it should not, because this change makes the Analyzer stateful.

Feel free to ignore it again.

Reply:
You should modify https://github.com/dudykr/stc/blob/e28c20fd9fc24940e47d7b8d319ce1847ba626c8/crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs#L3240 and https://github.com/dudykr/stc/blob/e28c20fd9fc24940e47d7b8d319ce1847ba626c8/crates/stc_ts_file_analyzer/src/analyzer/expr/call_new.rs#L3250

If it still fails when you change both, it means the latter is the case.

Reply:
It worked by adding `with_ctx` on the second match arm. Im still not 100% sure my solution is good enough so please take that into account.